channels_redis/alt.py

```diff
@@ -0,0 +1,311 @@
+            if channel in self.channels:
+                del self.channels[channel]
+                try:
+                    shard = self._get_shard(channel)
+                    await shard.unsubscribe(channel)
+                except Exception:
+                    traceback.print_exc(file=sys.stderr)
+                    # We don't re-raise here because we want the CancelledError to be the one re-raised.
+            raise
+
+        return msgpack.unpackb(message)
+
+    ################################################################################
+    # Groups extension
+    ################################################################################
+
+    async def group_add(self, group, channel):
+        """
+        Adds the channel name to a group.
+        """
+        if channel not in self.channels:
+            raise RuntimeError(
+                "You can only call group_add() on channels that exist in-process.\n"
+                "Consumers are encouraged to use the common pattern:\n"
+                f"   self.channel_layer.group_add({repr(group)}, self.channel_name)"
+            )
+        group_channel = self._get_group_channel_name(group)
+        if group_channel not in self.groups:
+            self.groups[group_channel] = set()
+        group_channels = self.groups[group_channel]
+        if channel not in group_channels:
+            group_channels.add(channel)
+        shard = self._get_shard(group_channel)
+        await shard.subscribe(group_channel)
+
+    async def group_discard(self, group, channel):
+        """
+        Removes the channel from a group.
+        """
+        group_channel = self._get_group_channel_name(group)
+        assert group_channel in self.groups
+        group_channels = self.groups[group_channel]
+        assert channel in group_channels
+        group_channels.remove(channel)
+        if len(group_channels) == 0:
+            del self.groups[group_channel]
+            shard = self._get_shard(group_channel)
+            await shard.unsubscribe(group_channel)
+
+    async def group_send(self, group, message):
+        """
+        Send the message to all subscribers of the group.
+        """
+        group_channel = self._get_group_channel_name(group)
+        shard = self._get_shard(group_channel)
+        await shard.publish(group_channel, message)
+
+
+def on_close_noop(sender, exc=None):
+    """
+    If you don't pass an `on_close` function to the `Receiver`, then it
+    defaults to one that closes the Receiver whenever the last subscriber
+    unsubscribes. That is not what we want; instead, we want the Receiver
+    to continue even if no one is subscribed, because soon someone *will*
+    subscribe and we want things to continue from there. Passing this
+    empty function solves it.
+    """
+    pass
+
+
+class RedisSingleShardConnection:
+    def __init__(self, host, channel_layer):
```

Comment:
initialize a redis.pubsub instance here.
We will be using the create_redis instance only for publishing.
use the internal client for subscribing.

Reply:
Well, we cannot connect to redis from `__init__` because we need a coroutine (`__init__` is synchronous; not a coroutine).

> use the internal client for subscribing

Not sure what you mean. You cannot publish and subscribe/listen from the same redis connection. Rather you need one connection for publishing, and another for subscribing/listening.

Reply:
>Not sure what you mean. You cannot publish and subscribe/listen from the same redis connection. Rather you need one connection for publishing, and another for subscribing/listening.

I get you. What I meant was, we have one single connection for pulishing.
Each consumer has their own connection which they use to subscribe.

Reply:
Again, one connection to Redis *per consumer* is not efficient. Imagine you had one million consumers connected to one thousand django server replicas. Now imagine the network I/O for that if you also had one million connections to the *same* Redis server. It would be WAY BETTER to have only one thousand connections to Redis (one per server replica) and each server Replica did some internal broadcasting in-memory.